x/warp/contract.go

```diff
@@ -99,34 +112,168 @@ func getBlockchainID(accessibleState contract.AccessibleState, caller common.Add
+	if err != nil {
+		return nil, remainingGas, fmt.Errorf("%w: %s", errInvalidPredicateBytes, err)
+	}
+	warpMessage, err := warp.ParseMessage(unpackedPredicateBytes)
+	if err != nil {
+		return nil, remainingGas, fmt.Errorf("%w: %s", errInvalidWarpMsg, err)
+	}
+
+	blockHashPayload, err := warpPayload.ParseBlockHashPayload(warpMessage.UnsignedMessage.Payload)
+	if err != nil {
+		return nil, remainingGas, fmt.Errorf("%w: %s", errInvalidBlockHashPayload, err)
+	}
+	packedOutput, err := PackGetVerifiedWarpBlockHashOutput(GetVerifiedWarpBlockHashOutput{
+		WarpBlockHash: WarpBlockHash{
+			OriginChainID: common.Hash(warpMessage.SourceChainID),
+			BlockHash:     blockHashPayload.BlockHash,
+		},
+		Valid: true,
+	})
+	if err != nil {
+		return nil, remainingGas, err
+	}
+
+	// Return the packed output and the remaining gas
+	return packedOutput, remainingGas, nil
+}
+
+// UnpackGetVerifiedWarpMessageInput attempts to unpack [input] into the *big.Int type argument
+// assumes that [input] does not include selector (omits first 4 func signature bytes)
+func UnpackGetVerifiedWarpMessageInput(input []byte) (*big.Int, error) {
+	res, err := WarpABI.UnpackInput("getVerifiedWarpMessage", input)
+	if err != nil {
+		return new(big.Int), err
+	}
+	unpacked := *abi.ConvertType(res[0], new(*big.Int)).(**big.Int)
+	return unpacked, nil
+}
+
+// PackGetVerifiedWarpMessage packs [index] of type *big.Int into the appropriate arguments for getVerifiedWarpMessage.
+// the packed bytes include selector (first 4 func signature bytes).
 // This function is mostly used for tests.
-func PackGetVerifiedWarpMessage() ([]byte, error) {
-	return WarpABI.Pack("getVerifiedWarpMessage")
+func PackGetVerifiedWarpMessage(index *big.Int) ([]byte, error) {
+	return WarpABI.Pack("getVerifiedWarpMessage", index)
 }
 
 // PackGetVerifiedWarpMessageOutput attempts to pack given [outputStruct] of type GetVerifiedWarpMessageOutput
 // to conform the ABI outputs.
 func PackGetVerifiedWarpMessageOutput(outputStruct GetVerifiedWarpMessageOutput) ([]byte, error) {
 	return WarpABI.PackOutput("getVerifiedWarpMessage",
 		outputStruct.Message,
-		outputStruct.Exists,
+		outputStruct.Valid,
 	)
 }
 
+// UnpackGetVerifiedWarpMessageOutput attempts to unpack [output] as GetVerifiedWarpMessageOutput
+// assumes that [output] does not include selector (omits first 4 func signature bytes)
+func UnpackGetVerifiedWarpMessageOutput(output []byte) (GetVerifiedWarpMessageOutput, error) {
+	outputStruct := GetVerifiedWarpMessageOutput{}
+	err := WarpABI.UnpackIntoInterface(&outputStruct, "getVerifiedWarpMessage", output)
+
+	return outputStruct, err
+}
+
+func unpackWarpMessage(accessibleState contract.AccessibleState, input []byte) ([]byte, bool, error) {
+	warpIndex, err := UnpackGetVerifiedWarpMessageInput(input)
+	if err != nil {
+		return nil, false, err
+	}
+	if !warpIndex.IsUint64() {
+		return nil, false, fmt.Errorf("invalid index: %v", warpIndex)
+	}
+	warpIndexInt := int(warpIndex.Uint64())
```

Comment:
is this conversion OK if it leads to a negative number here? I think we should check for IsInt64 and >= 0

Reply:
Will fix the conversion

Reply:
Nice catch, adding tests for maxuint64 and maxint64 as well